src/main/java/gregtech/api/metatileentity/multiblock/ui/MultiblockUIBuilder.java

```diff
@@ -0,0 +1,982 @@
+     */
+    public MultiblockUIBuilder addFuelNeededLine(String fuelName, int previousRecipeDuration) {
+        if (!isStructureFormed || !isActive) return this;
+        fuelName = getSyncer().syncString(fuelName);
+        previousRecipeDuration = getSyncer().syncInt(previousRecipeDuration);
+        addKey(KeyUtil.lang(TextFormatting.GRAY,
+                "gregtech.multiblock.turbine.fuel_needed",
+                KeyUtil.string(TextFormatting.RED, fuelName),
+                KeyUtil.number(TextFormatting.AQUA, previousRecipeDuration)));
+        return this;
+    }
+
+    /**
+     * Adds the name of a recipe map to the display.
+     *
+     * @param map the {@link RecipeMap} to get the name of
+     */
+    public MultiblockUIBuilder addRecipeMapLine(RecipeMap<?> map) {
+        if (!isStructureFormed) return this;
+
+        IKey mapName = KeyUtil.lang(TextFormatting.YELLOW, map.getTranslationKey());
+        addKey(KeyUtil.lang(TextFormatting.GRAY, "gregtech.multiblock.machine_mode", mapName));
+
+        return this;
+    }
+
+    /**
+     * Adds the current outputs of a recipe from recipe logic. Items then fluids.
+     *
+     * @param arl an instance of an {@link AbstractRecipeLogic} to gather the outputs from.
+     */
+    public MultiblockUIBuilder addRecipeOutputLine(@NotNull AbstractRecipeLogic arl) {
+        return addRecipeOutputLine(arl, DEFAULT_MAX_RECIPE_LINES);
+    }
+
+    /**
+     * Adds the current outputs of a recipe from recipe logic. Items then fluids.
+     *
+     * @param arl      an instance of an {@link AbstractRecipeLogic} to gather the outputs from.
+     * @param maxLines the maximum number of lines to print until truncating with {@code ...}
+     */
+    public MultiblockUIBuilder addRecipeOutputLine(AbstractRecipeLogic arl, int maxLines) {
+        // todo doing this every tick on the server is probably not good
+        Recipe recipe = arl.getParallelRecipe();
+
+        if (getSyncer().syncBoolean(recipe == null)) return this;
+        if (getSyncer().syncBoolean(arl.getRecipeMap() == null)) return this;
+
+        // noinspection DataFlowIssue
+        long eut = getSyncer().syncLong(() -> recipe.getEUt());
+        long maxVoltage = getSyncer().syncLong(arl.getMaximumOverclockVoltage());
+
+        List<ItemStack> itemOutputs = new ArrayList<>();
+        List<ChancedItemOutput> chancedItemOutputs = new ArrayList<>();
+        List<FluidStack> fluidOutputs = new ArrayList<>();
+        List<ChancedFluidOutput> chancedFluidOutputs = new ArrayList<>();
+
+        if (isServer()) {
+            // recipe is checked indirectly for null
+            // noinspection DataFlowIssue
+            itemOutputs.addAll(recipe.getOutputs());
+            chancedItemOutputs.addAll(recipe.getChancedOutputs().getChancedEntries());
+            fluidOutputs.addAll(recipe.getFluidOutputs());
+            chancedFluidOutputs.addAll(recipe.getChancedFluidOutputs().getChancedEntries());
```

Comment:
Will this list of chanced outputs be accurate here? Since the chanced outputs are not calculated in the call of `arl.getParallelRecipe()`, they are instead calculated in `arl#setupRecipe()`. 

Unless you want to do a list of all possible chanced outputs, not just the specific outputs that will result from the current recipe (Which is what I would prefer btw)

Reply:
the idea is to list all possible chanced outputs, as there could be concerns with people trying to rig the chances if it showed only the successful rolls

Reply:
Ah, ok. I was not sure if it was intended to be like that, or if we were showing what chanced outputs were actually going to occur. I prefer the way of showing all the chanced outputs, although we should make it clear that it is a list of all of them.

Something like a separate section from the normal outputs saying, "Possible Outputs:" and then listing the chanced output and the current chance (based on current voltage tier of the multiblock and chanced output scaling) or something like that, so it is clear to the player that they might not get them

Reply:
the listed chance amount should already be scaled with the amount of overclocks